### PR TITLE
Initial NEMU CI

### PR DESCRIPTION
**Pull request by sameo**
This is the first basic `semaphoreci` and `pullapprove` based CI for NEMU. We have:

* A `pullapprove` configuration file where 2 LGTMs from the nemu-write team are needed for a PR to be merged.
* A couple of scripts for running minimal CI in a `semaphoreci` context

### CODE REVIEW

**Comment by sameo** [comment]
@mcastelino @rbradford Please not that we need to get additional privileges in order to enable pullapprove and semaphore CI to gate our wip-master branch.

File: pc-bios/optionrom/code16gcc.h [hunk 1]
```diff
@@ -0,0 +1,3 @@
+asm(
```
Comment (mcastelino) [inline pc-bios/optionrom/code16gcc.h]: @sameo why was this needed?

File: tools/CI/minimal_ci.sh [hunk 2]
```diff
@@ -37,6 +37,7 @@ DOWNLOAD_ONLY="false"
 RUN_UNSAFE="false"
 VERBOSE="false"
 CHECK="false"
+VSOCK="true"
```
Comment (mcastelino) [inline tools/CI/minimal_ci.sh]: thanks. I should have done this before. This will also allow folks to test on older host kernels.

File: pc-bios/optionrom/code16gcc.h (hunk 1 above)
Comment (sameo) [reply pc-bios/optionrom/code16gcc.h]: The semaphoreci toolchain is gcc 4.8, and  from pc-bios/optionrom/Makefile:

```
ifeq ($(filter -m16, $(QEMU_CFLAGS)),)
# Attempt to work around compilers that lack -m16 (GCC <= 4.8, clang <= ??)
# On GCC we add -fno-toplevel-reorder to keep the order of asm blocks with
# respect to the rest of the code.  clang does not have -fno-toplevel-reorder,
# but it places all asm blocks at the beginning and we're relying on it for
# the option ROM header.  So just force clang not to use the integrated
# assembler, which doesn't support .code16gcc.
QEMU_CFLAGS += $(call cc-option, $(QEMU_CFLAGS), -fno-toplevel-reorder)
QEMU_CFLAGS += $(call cc-option, $(QEMU_CFLAGS), -no-integrated-as)
QEMU_CFLAGS += -m32 -include $(SRC_PATH)/pc-bios/optionrom/code16gcc.h
endif
```

**Comment by sameo** [comment]
@mcastelino It highly depends on the networking bandwidth, but with a poor one it can take up to 1h.
I will have a look at enabling generic cache for caching the workloads images across builds.
If that does not work, we'll have to strip it down to 2-3 workloads, yes.

File: tools/CI/semaphore/setup.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -e
+
+# -fno-semantic-interposition is not supported by gcc 4.8 (from semaphore)
+export EXTRA_CFLAGS=" -O3 -falign-functions=32 -D_FORTIFY_SOURCE=2 -fPIE"
+export SRCDIR=~/nemu
+
+sudo apt-get install -y libcap-ng-dev librbd-dev
+
+$SRCDIR/tools/build_x86_64.sh
```
[the remark this answers is not in the document]
Comment (sameo) [reply tools/CI/semaphore/setup.sh]: > It would also allow the testing of vsock as the ccloudvm scripts use bionic which runs with a kernel new enough to do vsock.

Are you suggesting to also run the actual CI scripts inside ccloudvm?

[the remark this answers is not in the document]
Comment (rbradford) [reply tools/CI/semaphore/setup.sh]: @sameo I run the CI inside ccloudvm myself. We've successfully run qemu inside ccloudvm inside semaphore for CIAO so I think it should work.

File: tools/CI/semaphore/setup.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+sudo apt-get update
+sudo apt-get install qemu xorriso -y
+sudo chmod ugo+rwx /dev/kvm
+
+echo ""
+echo "===== Building ccloudvm ====="
+echo ""
+
+go version
+go get github.com/intel/ccloudvm
```
Comment (rbradford) [inline tools/CI/semaphore/setup.sh]: @sameo if you do go get github.com/intel/ccloudvm/... then it will do the right thing with the packages removing the need for the line below.

**Comment by rbradford** [comment]
@sameo with the changes i've pushed building and cross building works

I've also also pushed a change to run the CI script inside the ccloudvm, let's see if that works.

**Comment by sameo** [comment]
@rbradford Thanks for fixing the ccloudvm based build.
It seems running the CI inside ccloudvm failed.
I cleaned up the branch (I had a debug commit there) and removed your last commit to double check if CI passes now:

https://semaphoreci.com/sameo/nemu/branches/topic-ci/builds/38

**Comment by rbradford** [comment]
@sameo I've fixed the PATH inside the script now so it should hopefully find ccloudvm to do the run command.

Irrespective of that it seems to timeout after 1 hour mostly from downloading all the files. Either we could reduce what we do (based off some of our recently discoveries) or maybe use semaphore caching: https://semaphoreci.com/docs/caching-between-builds.html

**Comment by sameo** [comment]
@rbradford Thanks a lot, LGTM.
